# Loops

Usamos loops quando precisamos fazer uma operação várias vezes.
Às vezes, nós sabemos quantas vezes vamos fazer essa operação.
Às vezes, nós não sabemos, e precisamos ficar repetindo até que alguma condição seja satisfeita.

No primeiro caso, usamos um **for loop**.  
No segundo caso, usamos um **while loop**.

## For Loop

```python
for ... in ...
```

Muitas vezes, o loop é feito em um **range** de números. Nesses casos, o comando `range` pode ser bem útil

In [71]:
for i in range(10):
    print(i)

0
1
2
3
4
5
6
7
8
9


Podemos usar um loop para rodar uma função várias vezes....

In [48]:
def quadrado(x):
    return(x ** 2)

In [49]:
for i in range(10):
    print(quadrado(i))

0
1
4
9
16
25
36
49
64
81


Também podemos usar um loop dentro de uma função

In [50]:
def pares_menores_que(n):
    for i in range(n):
        if i % 2 == 0:
            print(i)

In [51]:
pares_menores_que(12)

0
2
4
6
8
10


Note que `range(n)` não inclui o número `n`.
Se quisermos incluir `n`, temos que fazer `range(n + 1)`:

In [52]:
def soma_tudo_ateh(n):
    soma = 0
    for i in range(n + 1):
        soma = soma + i
    return(soma)

In [54]:
#0 + 1 + 2 + 3
soma_tudo_ateh(3)

6

In [56]:
soma_tudo_ateh(10**6)

500000500000

Já vimos que `range(n)` dá todos os inteiros de 0 até $n-1$.  
E se não quisermos começar a contagem do zero?  
E se quisermos contar de 2 em 2?  
Para isso, podemos passar 3 argumentos para `range`.

O comando `range(a, b, s)` conta todos os números de $a$ até $b-1$, pulando de $s$ em $s$.

In [60]:
for i in range(10, 20, 3):
    print(i)

10
13
16
19


In [63]:
def soma_impares_ate(n):
    soma = 0
    for i in range(1, n+1, 2): #so galera ímpar!
        soma = soma + i
    return(soma)

In [68]:
#1 + 3 + 5 + 7
soma_impares_ate(10)

25

In [69]:
def divisores(n):
    for d in range(1, n):
        if n % d == 0:
            print(d)

In [70]:
divisores(60)

1
2
3
4
5
6
10
12
15
20
30


## While loop

```python
i = 0
while  ...:
    faz_alguma_coisa
    i += 1
```

In [79]:
i = 0
while i <= 10:
    print(i)
    i = i + 1

0
1
2
3
4
5
6
7
8
9
10


Um exemplo: listar todos os quadrados menores que um número.
Eu não sei quantos quadrados existem, então não posso usar um _for loop_.
Mas posso usar um _while loop_...

In [82]:
def quadrados_menores_que(n):
    i = 0
    while i ** 2 < n:
        print(i**2)
        i += 1

In [83]:
quadrados_menores_que(100)

0
1
4
9
16
25
36
49
64
81


#### Cuidado!
Ao usar _while loops_, tome muito cuidado para não cair em um **loop infinito**

In [85]:
i = 0
while i <= 10:
    print(i)

0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0


0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0


0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0


0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0


KeyboardInterrupt: 

### Aplicação: Taxa Interna de Retorno de um investimento e o Método de Newton

Um empreendedor vai fazer um investimento que consiste em investir 100 milhões agora e obter 30 milhões nos 4 anos seguintes. Ele vai financiar esse investimento junto a um banco, que vai lhe cobrar uma taxa de juros. Qual é a maior taxa de juros que o empreendedor está disposto a pagar para, ainda assim, ter lucro com esse investimento?

Quando você estudar Matemática Financeira, você vai aprender que a resposta para essa pergunta é a raiz desta função:

$$
f(x) = -100 + \frac{30}{1+x} + \frac{30}{(1+x)^2} + \frac{30}{(1+x)^3} + \frac{30}{(1+x)^4}
$$

Como podemos encontrar a raiz dessa função?

Vocë vai aprender em Cálculo que a resposta se chama **Método de Newton**. A lógica é simples:

![metodo-newton](http://www.cs.utsa.edu/~wagner/CS3343/newton/newton.gif)

1. Chuta um valor qualquer para $x$.
2. Descubra a reta tangente à função naquele ponto e encontre a raiz dessa reta. Essa raiz é uma aproximação para a raiz da função e será o seu novo "chute".
3. Fique repetindo o passo 2 até que o valor que você está calculando mude muito pouquinho. Esse valor será, aproximadamente, o valor da raiz que você está buscando.

Você vai aprender em Cálculo que a raiz da reta tangente a uma função $f(x)$ em um ponto $x_0$ é

$$
x_1 = x_0 - \frac{f(x)}{f'(x)}
$$

onde 

$$
f'(x) = \frac{dy}{dx} = \lim_{dx \to 0} \frac{f(x + dx) - f(x)}{dx}
$$

é chamada **derivada de f(x) no ponto x**.

Calma. Você vai entender isso melhor quando estudar Cálculo.  
O objetivo aqui é mostrar como é fácil programar isso em Python.   

Vamos começar programando uma função que dá a derivada de uma outra função. Como a derivada implica um $dx$ pequenininho, vamos tomar um $dx$ de 0.0001.

In [ ]:
def derivada(f, x, dx = 0.0001):
    dy = f(x + dx) - f(x)
    return(dy/dx)

Agora, vamos criar uma função que, dada uma função $f(x)$ e um ponto $x$, calcula a raiz da reta tangente à função nesse ponto. Essa função meramente aplica a fórmula

$$
x_1 = x_0 - \frac{f(x)}{f'(x)}
$$


In [ ]:
def raiz_reta_tangente(f, x):
    return(x - f(x)/derivada(f, x))

O método de Newton consiste em aplicar a função acima várias vezes, enquanto o valor de $x_1$ e $x_0$ não mudarem mais. Podemos fazer isso usando o tema da aula de hoje: um _while loop_.

In [67]:
def raiz(f, x0, tol = 0.0001):
    
    x1 = raiz_reta_tangente(f, x0)
    
    while abs(x1 - x0) > tol:
        x0 = x1
        x1 = raiz_reta_tangente(f, x0)
    
    return(x1)

Vamos testar se nossa função funciona na função $f(x) = x^2 - 25$. Sabemos que a raiz é 5. Logo, se nossas funções estiverem funcionando, o resultado deverá ser 7.

Primeiro, montamos a função $f(x) = x^2 - 25$

In [77]:
def f(x):
    return(x**2 - 25)

Agora, verificamos que nossa função funciona. Precisamos de um chute inicial. Vou colocar o dia do meu aniversário: 1301, um número bem, bem longe da raiz correta.

In [78]:
raiz(f, x0 = 1301)

5.000000000015194

Na mosca!!!! Vamos agora voltar ao problema do empreendedor. Qual é a maior taxa de juros que ele estaria disposo a pagar ao banco para, ainda assim, ter lucro no seu investimento?

In [79]:
def f(x):
    return(-100 + 30/(1+x) + 30/(1+x)**2 + 30/(1+x)**3 + 30/(1+x)**4)

In [80]:
raiz(f, 0)

0.07713847295836763

Temos a resposta: 7.7% é o maior valor que um empreendedor estaria disposto a pagar para financiar esse empreendimento. Esse valor se chama **Taxa Interna de Retorno** do empreendimento, e é um dos valores mais importantes para avaliar se um investimento vale a pena.